README
NAME
    Alien::SDL - building, finding and using SDL binaries

VERSION
    Version 1.444

SYNOPSIS
    Alien::SDL tries (in given order) during its installation:

    * When given `--with-sdl-config' option use specified sdl-config script
    to locate SDL libs.
         perl Build.PL --with-sdl-config=/opt/sdl/bin/sdl-config

        or using default script name 'sdl-config' by running:

         perl Build.PL --with-sdl-config

        IMPORTANT NOTE: Using --with-sdl-config avoids considering any other
        build methods; no prompt with other available build options.

    * Locate an already installed SDL via 'sdl-config' script.
    * Check for SDL libs in directory specified by SDL_INST_DIR variable. In
    this case the module performs SDL library detection via
    '$SDL_INST_DIR/bin/sdl-config' script.
         SDL_INST_DIR=/opt/sdl perl ./Build.PL

    * Download prebuilt SDL binaries (if available for your platform).
    * Build SDL binaries from source codes (if possible on your system).

    Later you can use Alien::SDL in your module that needs to link agains
    SDL and/or related libraries like this:

        # Sample Makefile.pl
        use ExtUtils::MakeMaker;
        use Alien::SDL;

        WriteMakefile(
          NAME         => 'Any::SDL::Module',
          VERSION_FROM => 'lib/Any/SDL/Module.pm',
          LIBS         => Alien::SDL->config('libs', [-lAdd_Lib]),
          INC          => Alien::SDL->config('cflags'),
          # + additional params
        );

DESCRIPTION
    Please see Alien for the manifesto of the Alien namespace.

    In short `Alien::SDL' can be used to detect and get configuration
    settings from an installed SDL and related libraries. Based on your
    platform it offers the possibility to download and install prebuilt
    binaries or to build SDL & co. from source codes.

    The important facts:

    * The module does not modify in any way the already existing SDL
    installation on your system.
    * If you reinstall SDL libs on your system you do not need to reinstall
    Alien::SDL (providing that you use the same directory for the new
    installation).
    * The prebuild binaries and/or binaries built from sources are always
    installed into perl module's 'share' directory.
    * If you use prebuild binaries and/or binaries built from sources it
    happens that some of the dynamic libraries (*.so, *.dll) will not
    automaticly loadable as they will be stored somewhere under perl
    module's 'share' directory. To handle this scenario Alien::SDL offers
    some special functionality (see below).

METHODS
  config()
    This function is the main public interface to this module. Basic
    functionality works in a very similar maner to 'sdl-config' script:

        Alien::SDL->config('prefix');   # gives the same string as 'sdl-config --prefix'
        Alien::SDL->config('version');  # gives the same string as 'sdl-config --version'
        Alien::SDL->config('libs');     # gives the same string as 'sdl-config --libs'
        Alien::SDL->config('cflags');   # gives the same string as 'sdl-config --cflags'

    On top of that this function supports special parameters:

        Alien::SDL->config('ld_shared_libs');

    Returns a list of full paths to shared libraries (*.so, *.dll) that will
    be required for running the resulting binaries you have linked with SDL
    libs.

        Alien::SDL->config('ld_paths');

    Returns a list of full paths to directories with shared libraries (*.so,
    *.dll) that will be required for running the resulting binaries you have
    linked with SDL libs.

        Alien::SDL->config('ld_shlib_map');

    Returns a reference to hash of value pairs '<libnick>' =>
    '<full_path_to_shlib'>, where '<libnick>' is shortname for SDL related
    library like: SDL, SDL_gfx, SDL_net, SDL_sound ... + some non-SDL
    shortnames e.g. smpeg, jpeg, png.

    NOTE: config('ld_<something>') return an empty list/hash if you have
    decided to use SDL libraries already installed on your system. This
    concerns 'sdl-config' detection and detection via
    '$SDL_INST_DIR/bin/sdl-config'.

  check_header()
    This function checks the availability of given header(s) when using
    compiler options provided by "Alien::SDL->config('cflags')".

        Alien::SDL->check_header('SDL.h');
        Alien::SDL->check_header('SDL.h', 'SDL_net.h');

    Returns 1 if all given headers are available, 0 otherwise.

  get_header_version()
    Tries to find a header file specified as a param in SDL prefix direcotry
    and based on "#define" macros inside this header file tries to get a
    version triplet.

        Alien::SDL->get_header_version('SDL_mixer.h');
        Alien::SDL->get_header_version('SDL_version.h');
        Alien::SDL->get_header_version('SDL_gfxPrimitives.h');
        Alien::SDL->get_header_version('SDL_image.h');
        Alien::SDL->get_header_version('SDL_mixer.h');
        Alien::SDL->get_header_version('SDL_net.h');
        Alien::SDL->get_header_version('SDL_ttf.h');
        Alien::SDL->get_header_version('smpeg.h');

    Returns string like '1.2.3' or undef if not able to find and parse
    version info.

BUGS
    Please post issues and bugs at
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-SDL

AUTHOR
        Kartik Thakore
        CPAN ID: KTHAKORE
        Thakore.Kartik@gmail.com
        http://yapgh.blogspot.com

ACKNOWLEDGEMENTS
        kmx - complete redesign between versions 0.7.x and 0.8.x

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl module Alien::SDL

Versioning rule: public releases are even numbers, dev releases are odd. (same like perl dist)

TODO
     - fixed spelling and grammar in docs
     - check for available libs when using --with-sdl-config option

1.445_2 Sun Feb 15 2015
    - Apply patch from Debian [jtpalmer]

1.445_1 Thu Aug 14 2014
     - fix build on OS X Mavericks (missing symbol CGDirectPaletteRef) [FROGGS]

1.444 Wed May 14 2014
     - check that perl was linked against pthread and do a pthread-less build if not (OpenBSD) [FROGGS]
     - add patch that replaces $^ in Makefile.in, which can be a problem for BSD's make [FROGGS]
     - let SDL_Pango.h include ft2build.h, because even when the header file ifdef's out some
       code that needs it, the source file does not and thus explodes happily. [FROGGS]
     - add PIC flags to CFLAGS when the compiler can handle it, this is meant to fix
       an issue on freebsd [FROGGS]

1.442 Tue Apr 22 2014
     - bump SDL_gfx to 2.0.25 [FROGGS]
     - fix for building on >=libX11-1.5.99.901 [FROGGS]

1.440 Sun Aug 12 2012
     - patch for File::Temp->tempfile -> File::Temp::tempfile (#244) [FROGGS]

1.439_1 Sun Aug 12 2012
     - added patch for SDL::GFX::ImageFilter::MMX_detect (by Steve Langasek, ubuntu bug 1029981) [FROGGS]
     - source code build option is now only available when SDL itself can/will be built [FROGGS]

1.438 Sun Aug 12 2012
     - added minimum Perl version [jtpalmer]
     - added release tests to check pod syntax and coverage [jtpalmer]
     - return ld_shlib_map, ld_shared_libs and ld_paths from config when using script option [jtpalmer]
     - excluding undef paths from ld_shlib_map and ld_shared_libs [jtpalmer]
     - changed shlib regex to match libz and libfreetype [jtpalmer]
     - disabling mmx for SDL_gfx for armv6-unix [jtpalmer]
     - added bin script to make configuring easier [kthakore]
     - added CGFloat patch for darwin/SDL_image [FROGGS]
     - CFLAGS: added several circumstances where we need /usr/X11R7/... [FROGGS]
     - CFLAGS: removed /usr/pkg/..., that clashed on my netbsd [FROGGS]
     - CFLAGS: removed duplicate /usr/X11R6/... [FROGGS]
     - setting LD_LIBRARY_PATH for all BSDs [FROGGS]
     - added missing space in cflags and ldflags [FROGGS]
     - adding rpath linker flag for linux and bsd [FROGGS]

1.436 Tue Jun 19 2012
     - reapplied openbsd-rldflags patch, it got lost during update to SDL-1.2.15 [FROGGS]
     - fixed SDL dirname during Build when applying patches (was SDL-1.2.14, is SDL-1.2) [FROGGS]
     - removed cl from cc_re for gcc binary sets, made new binary set for ActivePerl instead [FROGGS]
     - checking environment on windows when compiler is cl [FROGGS]
     - using vcvars32.bat to set up env on windows [FROGGS]
     - cl compiler prints to stdout instead of stderr, displaying these messages [FROGGS]
     - added libs like ogg to shlib_map, needed for loading these libs at runtime to help msvc [FROGGS]
     - added patches/SDL_config_win32.h for msvc, because it has no configure [FROGGS]
     - setting ld_shlib_map for 'already installed'-option too [FROGGS]
     - msvc binary package contains patched SDL_rwops.h [FROGGS]

1.434 Thu May 31 2012
     - added compiler checks [FROGGS]

1.432 Wed May 30 2012
     - added cl to cc_re for binary sets, this way ActivePerl works again [FROGGS]
     - configure-like appearance, first check, then do [FROGGS]
     - libSDL is newest one from hg-repo, this should fix the _XGetRequest issue [FROGGS]
     - hide proxy passwords [FROGGS]
     - build libogg/libvorbis if its not there [FROGGS]
     - fixed logic for detecting proxies [FROGGS]
     - added libs like libSDL 1.2.15, tweaked patches [FROGGS]
     - fixed bugtracker link [FROGGS]

1.430 Wed Jan 04 2012
     - readding source build sets that got lost [FROGGS]
     - adding extra PATH for make command like we do for configure on solaris [FROGGS]
     - setting LD_LIBRARY_PATH for GNU/kFreeBSD too so libs like SDL_image will find libSDL [FROGGS]
     - new Win/64bit binaries (from strawberryperl 5.14+) + gccversion detection magic [KMX]
     - new Win/32bit binaries (the same lib versions as 64bit, set as "secondary" option) [KMX]

1.428 Wed Jul 20 2011
     - added Capture::Tiny to build_requires [FROGGS]
     - added search paths for command ar on solaris [FROGGS]
     - 'movq' bugfix: http://hg.libsdl.org/SDL/rev/32f0f603a0c8 [FROGGS]
     - added 'RELEASE' to get_header_version (e.g. png.h has it) [FROGGS]
     - libpng15 bugfix: https://bugs.gentoo.org/show_bug.cgi?id=354675 [FROGGS]
     - added support for (authenticating) proxies when running from within CPAN shell [FROGGS]

1.427 Sun Jun 18 2011
     - SDL_Pango needs newer libtool to process ld scripts (like on ubuntu), patch added [FROGGS]
     - using `uname -a` to detect sparc systems to disable MMX [FROGGS]

1.426 Sun Apr 24 2011
     - added '--disable-sdltest' to SDL_mixer|ttf|gfx|Pango's configure for darwin [FROGGS]
     - fixing 'Use of uninitialized value ..' warnings in ACTION_install [kmx]
     - disabled MMX for SDL_gfx on ARMv5 machines [FROGGS]

1.425 Thu Feb 03 2011
     - setting --disable-video-x11-xv and -DNO_SHARED_MEMORY when XShm.h is missing #221 [FROGGS]

1.424 Wed Feb 02 2011
     - adding '--disable-sdltest' to SDL_image's configure for darwin #222 [FROGGS]
     - added patch for SDL_image/libpng-flags for netbsd #232 [FROGGS]

1.423 Tue Jan 25 2011
     - now '--with-sdl-config' is the same as '--with-sdl-config=sdl-config' [kmx]
     - added Build.PL option --with-sdl-config=... [madsen,kmx]

1.422 Wed Jan 19 2011
     - the same as latest dev release
     - added paths /usr/pkg/include and /usr/pkg/lib for dragonfly [FROGGS]
     - patch for SDL's configure script, -Wl,-rpath,LIBDIR for openbsd [FROGGS]
     - updated check_header() to use gcc on cygwin [FROGGS]
     - building libjpeg-v8b[FROGGS]
     - setting cc to gcc on cygwin [FROGGS]

1.421  Sat Oct 30 2010

Build.PL
use strict;
use warnings;

use lib "inc";
use File::Spec::Functions qw(catdir catfile);
use Config;
use My::Utility qw(check_config_script check_prebuilt_binaries check_prereqs_libs check_prereqs_tools $source_packs
                   check_perl_buildlibs);

my ( $ans, $travis ) = 0;

print "Welcome to Alien::SDL module installation\n";
print "-----------------------------------------\n";
print "checking operating system... $^O\n";
$| = 1;
print "checking for $My::Utility::cc... ";
if( !check_prereqs_tools($My::Utility::cc) ) {
  print "no\n";
  for(qw(gcc cc lc)) {
    next if $_ eq \$My::Utility::cc;
    print "checking for $_... ";
    if( check_prereqs_tools($_) ) {
      $My::Utility::cc = $_;
      print "yes\n";
      last;
    }
    else {
      print "no\n";
    }
  }
}
else {
  print "yes\n";
}
print "checking build system type... $Config{archname}\n";

#### we need the platform-specific module
my %platforms =(
       # Unix = default, thus not listing all UNIX like systems
       MSWin32 => 'Windows',
);
my $package = 'My::Builder::' . ($platforms{$^O} || 'Unix');
print "checking platform specific module... using '$package'\n";
eval "require $package" or die "Require '$package' failed: $@\n";

my $sdl_config;

#### Stadard Module::Builder stuff
my $build = $package->new(
  module_name          => 'Alien::SDL',
  all_from             => 'lib/Alien/SDL.pm',
  dist_abstract        => 'Get, Build and Use SDL libraries',
  dist_author          => 'Kartik Thakore <KTHAKORE@cpan.org>',
  license              => 'perl',
  requires             => {
    'File::Spec'       => '0',
    'File::Temp'       => '0',
    'File::ShareDir'   => '0',
    'File::Which'      => '0',
    'ExtUtils::CBuilder' => '0',
    'Capture::Tiny'    => '0',
    'perl'             => '5.008000',
  },
  build_requires       => {  #need to have for running: ./Build (install|test)
    'File::Spec'       => '0',
    'File::Temp'       => '0',
    'File::ShareDir'   => '0',
    'ExtUtils::CBuilder' => '0',
    'File::Path'       => '2.08',
    'File::Fetch'      => '0.24',
    'File::Find'       => '0',
    'File::Which'      => '0',
    'Digest::SHA'      => '0',
    'Archive::Extract' => '0',
    'Archive::Tar'     => '0',
    'Archive::Zip'     => '0',
    'Module::Build'    => '0.36',
    'Text::Patch'      => '1.4',
  },
  configure_requires   => {  #need to have for running: perl Build.PL
    'File::Spec'       => '0',
    'File::Path'       => '2.08',
    'File::Fetch'      => '0.24',
    'File::Find'       => '0',
    'File::Which'      => '0',
    'Digest::SHA'      => '0',
    'Archive::Extract' => '0',
    'Module::Build'    => '0.36',
    'Text::Patch'      => '1.4',
    'File::ShareDir'   => '0',
    'Capture::Tiny'    => '0',
  },
  meta_merge => {
    resources  => {
      bugtracker => 'http://github.com/PerlGameDev/SDL/issues?labels=Alien-SDL',
      repository => 'http://github.com/PerlGameDev/Alien-SDL'
    }
  },
  get_options => { 
      'with-sdl-config' => { qw(type :s  store) => \$sdl_config },
      'travis'          => { store => \$travis },
  },
  dynamic_config => 1,
  create_readme => 1,
  share_dir => 'sharedir',
  # sharedir is used for storing compiled/prebuilt binaries of SDL lib + related libraries
  # avoid using 'share' name as M::B doe not handle well paths like /xx/yy/share/zz/ww/share/xx
);

my $choice;
my %have_libs = ();
my %perl_libs = ();

if (defined $sdl_config) {
  # handle --with-sdl-config (without params)
  $sdl_config = 'sdl-config' if $sdl_config eq '';
  # Don't prompt; just use specified location:
  $choice = check_config_script($sdl_config)
      or warn "###ERROR### Unable to use config script $sdl_config\n";
}
else {
  $| = 1;

  if( $My::Utility::cc eq 'cl' && $^O eq 'MSWin32' ) {
    print "checking INCLUDE and LIB... ";
    if( !$ENV{INCLUDE} || !$ENV{LIB} ) {
      my @set = `\@vcvars32 & set`;
      chomp @set;
      my %set  = map /(\w+)=(.+)/, @set;
      for( keys %set ) {
        if( /^INCLUDE|LIB$/ ) {
          $ENV{$_} = $set{$_};
        }
      }
      print(($ENV{INCLUDE} && $ENV{LIB})
        ? "yes, via vcvars32\n"
        : "no\n");
    }
    else {
      print "yes\n";
    }
  }
  #### check what options we have for our platform
  my $rv;
  my @candidates = ();

  print "checking SDL_INST_DIR env var... ";
  if(defined($ENV{SDL_INST_DIR})) {
    if (-d $ENV{SDL_INST_DIR}) {
      print "yes, $ENV{SDL_INST_DIR}\n";
      my @sdlinst = File::Spec->splitdir($ENV{SDL_INST_DIR});
      if($rv = check_config_script(File::Spec->catdir(@sdlinst, 'bin', 'sdl-config'))) {
        push @candidates, $rv;
      }
      elsif($rv = check_config_script(File::Spec->catdir(@sdlinst, 'sdl-config'))) {
        push @candidates, $rv;
      }
    }
    else {
      print "directory '$ENV{SDL_INST_DIR}' does not exist";
    }
  }
  print "no\n";

  # sdl-config script
  push @candidates, $rv if $rv = check_config_script("sdl-config");

  if( $build->can_build_binaries_from_sources || scalar(@candidates) ) {
    for(qw(pthread SDL
           z jpeg tiff png SDL_image
           ogg vorbis vorbisfile SDL_mixer
           freetype SDL_ttf
           SDL_gfx
           pangoft2 pango gobject gmodule glib fontconfig expat SDL_Pango)) {
      $have_libs{$_} = check_prereqs_libs($_);
    }
  }

  $perl_libs{pthread} = check_perl_buildlibs('pthread') if $have_libs{pthread} && $^O eq 'openbsd';

  # prebuilt binaries (windows only)
  push @candidates, @{$rv} if $rv = check_prebuilt_binaries($build->os_type);

  if($build->can_build_binaries_from_sources) {
    for my $p ( @$source_packs ) {
      $rv           = { title => $p->{title}, members => [], buildtype => 'build_from_sources' };
      my $build_SDL = 0;
      for my $m (@{ $p->{members} }) {
        next if $m->{pack} !~ /^SDL/ && $have_libs{ $m->{pack} }[0];
        my $good = 1;
        $good   &= $have_libs{$_} && $have_libs{$_}[0] ? 1 : 0 for @{ $m->{prereqs}->{libs} };
        if( $good ) {
          $have_libs{ $m->{pack} }[0] ||= 1;
          $build_SDL                    = 1 if $m->{pack} eq 'SDL';
          push @{ $rv->{members} }, $m;
          $rv->{title} .= "$m->{pack}(v$m->{version}) ";
        }
      }
      push @candidates, $rv if scalar( @{ $rv->{members} } ) && $build_SDL;
    }
  };

  push @candidates, { title => 'Quit installation', buildtype => '' };

  #### ask user what way to go
  my $i                     = 1;
  my $prompt_string         = "\nYou have the following options:\n";
  my $recommended_candidate = 1;
  foreach my $c (@candidates) {
    $recommended_candidate = $i if $c->{buildtype} eq 'build_from_sources';

    if( $c->{buildtype} eq 'use_config_script' ) {
      $c->{title} .= "\n    ";
      for(qw(SDL SDL_image SDL_mixer SDL_ttf SDL_gfx SDL_Pango)) {
        $c->{title} .= "$_(v$have_libs{$_}->[0]) " if $have_libs{$_}[0];
      }
    }

    $prompt_string .= "[" . $i++ . "] " . $c->{title} . "\n";
  }

  # select option '1' for travis
  if ( defined $travis and $travis == 1 ) {
    $ans = 1;
  }

# or prompt user for build option
  else {
    $prompt_string .= "\nWhat way do you wanna go?";
    $ans = $build->prompt( $prompt_string, $recommended_candidate );
  }

  if($ans > 0 && $ans < scalar(@candidates)) {
    $choice = $candidates[$ans - 1];
  }

  $| = 0;
} # end else search and prompt for build method

#### store build params into 'notes'
if($choice) {
  print "Using \l$choice->{title}\n";
  $build->notes('build_params', $choice);
  $build->notes('env_include', $ENV{INCLUDE}) if $ENV{INCLUDE};
  $build->notes('env_lib',     $ENV{LIB})     if $ENV{LIB};
  $build->notes('have_libs',   \%have_libs);
  $build->notes('perl_libs',   \%perl_libs);
  $build->create_build_script();

  #### clean build_done stamp; force rebuild when running 'Build'
  $build->clean_build_done_marker;
}
else {
  $build->notes('build_params', undef); # just to be sure
  exit(0); # we want no reports from CPAN Testers in this case
}


inc/My/Builder/Windows.pm
package My::Builder::Windows;

use strict;
use warnings;
use base 'My::Builder';

sub can_build_binaries_from_sources {
  my $self = shift;
  return 0; # no
}

sub build_binaries {
  my( $self, $build_out, $build_src ) = @_;
  die "###ERROR### Building from sources not supported on MS Windows platform";
}

sub get_path {
  my ( $self, $path ) = @_;
  $path = '"' . $path . '"';
  return $path;
}

sub get_additional_cflags {
  my $self = shift;
  if($My::Utility::cc eq 'cl' && $self->notes('env_include')) {
    my $include = $self->notes('env_include');
    $include    =~ s/"//g;
    my @include = split(/;/, $include);
    my $cflags  = '';
    my $inc = $_;
    for( @include ) {
      my $inc = eval { require Win32; Win32::GetShortPathName($_); };
      $inc ||= $_;
      $cflags    .= "-I\"$inc\" " ;
    }
    return $cflags;
  }
  return '';
}

sub get_additional_libs {
  my $self = shift;
  if($My::Utility::cc eq 'cl' && $self->notes('env_lib')) {
    my $lib  = $self->notes('env_lib');
    $lib     =~ s/"//g;
    my @libs = split(/;/, $lib);
    my $libs = '';
    my $inc  = $_;
    for( @libs ) {
      my $_lib = eval { require Win32; Win32::GetShortPathName($_); };
      $_lib ||= $_;
      $libs  .= "/LIBPATH:\"$_lib\" " ;
    }
    return $libs;
  }
  return '';
}

1;


lib/Alien/SDL.pm
package Alien::SDL;
use strict;
use warnings;
use Alien::SDL::ConfigData;
use File::ShareDir qw(dist_dir);
use File::Spec;
use File::Find;
use File::Spec::Functions qw(catdir catfile rel2abs);
use File::Temp;
use Capture::Tiny;
use Config;

=head1 NAME

Alien::SDL - building, finding and using SDL binaries

=head1 VERSION

Version 1.445_2

=cut

our $VERSION = '1.445_2';
$VERSION = eval $VERSION;

=head1 SYNOPSIS

Alien::SDL tries (in given order) during its installation:

=over

=item * When given C<--with-sdl-config> option use specified sdl-config
script to locate SDL libs.

 perl Build.PL --with-sdl-config=/opt/sdl/bin/sdl-config

or use the default script named 'sdl-config' by running:

 perl Build.PL --with-sdl-config

B<IMPORTANT NOTE:> Using --with-sdl-config avoids considering any other
build methods; no prompt with other available build options.

=item * Locate an already installed SDL via 'sdl-config' script.

=item * Check for SDL libs in directory specified by SDL_INST_DIR variable.
In this case the module performs SDL library detection via
'$SDL_INST_DIR/bin/sdl-config' script.

 SDL_INST_DIR=/opt/sdl perl ./Build.PL

=item * Download prebuilt SDL binaries (if available for your platform).

=item * Build SDL binaries from source codes (if possible on your system).

=back

Later you can use Alien::SDL in your module that needs to link against SDL
and/or related libraries like this:

    # Sample Makefile.pl
    use ExtUtils::MakeMaker;
    use Alien::SDL;

    WriteMakefile(
      NAME         => 'Any::SDL::Module',
      VERSION_FROM => 'lib/Any/SDL/Module.pm',
      LIBS         => Alien::SDL->config('libs', [-lAdd_Lib]),
      INC          => Alien::SDL->config('cflags'),
      # + additional params
    );

=head1 DESCRIPTION

Please see L<Alien> for the manifesto of the Alien namespace.

In short C<Alien::SDL> can be used to detect and get
configuration settings from an installed SDL and related libraries.
Based on your platform it offers the possibility to download and
install prebuilt binaries or build SDL & co. from source codes.

The important facts:

=over

=item * The module does not modify in any way the already existing SDL
installation on your system.

=item * If you reinstall SDL libs on your system you do not need to
reinstall Alien::SDL (providing that you use the same directory for
the new installation).

=item * The prebuilt binaries and/or binaries built from sources are always
installed into perl module's 'share' directory.

=item * If you use the prebuilt binaries and/or binaries built from sources
it happens that some of the dynamic libraries (*.so, *.dll) will not
be automatically loadable as they will be stored somewhere under the perl
module's 'share' directory. To handle this scenario Alien::SDL offers some
special functionality (see below).

=back

=head1 METHODS

=head2 config()

This function is the main public interface to this module. Basic
functionality works in a very similar maner to 'sdl-config' script:

    Alien::SDL->config('prefix');   # gives the same string as 'sdl-config --prefix'
    Alien::SDL->config('version');  # gives the same string as 'sdl-config --version'
    Alien::SDL->config('libs');     # gives the same string as 'sdl-config --libs'
    Alien::SDL->config('cflags');   # gives the same string as 'sdl-config --cflags'

On top of that this function supports special parameters:

    Alien::SDL->config('ld_shared_libs');

Returns a list of full paths to shared libraries (*.so, *.dll) that will be
required for running the resulting binaries you have linked with SDL libs.

    Alien::SDL->config('ld_paths');

Returns a list of full paths to directories with shared libraries (*.so, *.dll)
that will be required for running the resulting binaries you have linked with
SDL libs.

    Alien::SDL->config('ld_shlib_map');

Returns a reference to hash of value pairs '<libnick>' => '<full_path_to_shlib'>,
where '<libnick>' is shortname for SDL related library like: SDL, SDL_gfx, SDL_net,
SDL_sound ... + some non-SDL shortnames e.g. smpeg, jpeg, png.

NOTE: config('ld_<something>') return an empty list/hash if you have decided to
use SDL libraries already installed on your system. This concerns 'sdl-config'
detection and detection via '$SDL_INST_DIR/bin/sdl-config'.

=head2 check_header()

This function checks the availability of given header(s) when using compiler
options provided by "Alien::SDL->config('cflags')".

    Alien::SDL->check_header('SDL.h');
    Alien::SDL->check_header('SDL.h', 'SDL_net.h');

Returns 1 if all given headers are available, 0 otherwise.

=head2 get_header_version()

Tries to find a header file specified as a param in SDL prefix direcotry and
based on "#define" macros inside this header file tries to get a version triplet.

    Alien::SDL->get_header_version('SDL_mixer.h');
    Alien::SDL->get_header_version('SDL_version.h');
    Alien::SDL->get_header_version('SDL_gfxPrimitives.h');
    Alien::SDL->get_header_version('SDL_image.h');
    Alien::SDL->get_header_version('SDL_mixer.h');
    Alien::SDL->get_header_version('SDL_net.h');
    Alien::SDL->get_header_version('SDL_ttf.h');
    Alien::SDL->get_header_version('smpeg.h');

Returns string like '1.2.3' or undef if not able to find and parse version info.

=head1 BUGS

Please post issues and bugs at L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-SDL>

=head1 AUTHOR

    Kartik Thakore
    CPAN ID: KTHAKORE
    Thakore.Kartik@gmail.com
    http://yapgh.blogspot.com

=head1 ACKNOWLEDGEMENTS

    kmx - complete redesign between versions 0.7.x and 0.8.x

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut

### get config params
sub config
{
  my $package = shift;
  my @params  = @_;
  return _sdl_config_via_script(@params)      if(Alien::SDL::ConfigData->config('script'));
  return _sdl_config_via_config_data(@params) if(Alien::SDL::ConfigData->config('config'));
}

### get version info from given header file
sub get_header_version {
  my ($package, $header) = @_;
  return unless $header;

  # try to find header
  my $root = Alien::SDL->config('prefix');
  #warn 'Finding in '.$root.'/include';
  my $include = File::Spec->catfile($root, 'include');
  my @files;
  find({ wanted => sub { push @files, rel2abs($_) if /\Q$header\E$/ }, follow => 1, no_chdir => 1, follow_skip => 2 }, $include);
  return unless @files;

  # get version info
  open(DAT, $files[0]) || return;
  my @raw=<DAT>;
  close(DAT);

  # generic magic how to get version major/minor/patchlevel
  my ($v_maj) = grep(/^#define[ \t]+[A-Z_]+?MAJOR[A-Z_]*[ \t]+[0-9]+/, @raw);
  $v_maj =~ s/^#define[ \t]+[A-Z_]+[ \t]+([0-9]+)[.\r\n]*$/$1/;
  my ($v_min) = grep(/^#define[ \t]+[A-Z_]+MINOR[A-Z_]*[ \t]+[0-9]+/, @raw);
  $v_min =~ s/^#define[ \t]+[A-Z_]+[ \t]+([0-9]+)[.\r\n]*$/$1/;
  my ($v_pat) = grep(/^#define[ \t]+[A-Z_]+(PATCHLEVEL|MICRO|RELEASE)[A-Z_]*[ \t]+[0-9]+/, @raw);
  $v_pat =~ s/^#define[ \t]+[A-Z_]+[ \t]+([0-9]+)[.\r\n]*$/$1/;
  return if (($v_maj eq '')||($v_min eq '')||($v_pat eq ''));
  return "$v_maj.$v_min.$v_pat";
}

### check presence of header(s) specified as params
sub check_header {
  my ($package, @header) = @_;
  print STDERR "[$package] Testing header(s): " . join(', ', @header);

  require ExtUtils::CBuilder; # PAR packer workaround

  my $config = {};
  if($^O eq 'cygwin') {
    my $ccflags = $Config{ccflags};
    $ccflags    =~ s/-fstack-protector//;
    $config     = { ld => 'gcc', cc => 'gcc', ccflags => $ccflags };
  }

  my $cb = ExtUtils::CBuilder->new( quiet => 1, config => $config );
  my ($fs, $src) = File::Temp::tempfile('aaXXXX', SUFFIX => '.c', UNLINK => 1);
  my $inc = '';
  my $i = 0;
  foreach (@header) {
    @header = (splice(@header, 0, $i) , 'stdio.h', splice(@header, $i)) if $_ eq 'jpeglib.h';
    $i++;
  }
  $inc .= "#include <$_>\n" for @header;
  syswrite($fs, <<MARKER); # write test source code
#if defined(_WIN32) && !defined(__CYGWIN__)
/* GL/gl.h on Win32 requires windows.h being included before */
#include <windows.h>
#endif
$inc
int demofunc(void) { return 0; }

MARKER
  close($fs);
  my $obj;
  my $stdout = '';
  my $stderr = '';
  ($stdout, $stderr) = Capture::Tiny::capture {
    $obj = eval { $cb->compile( source => $src, extra_compiler_flags => Alien::SDL->config('cflags')); };
  };

  if($obj) {
    print STDERR "\n";
    unlink $obj;
    return 1;
  }
  else {
    if( $stderr ) {
      $stderr =~ s/[\r\n]$//;
      $stderr =~ s/^\Q$src\E[\d\s:]*//;
      print STDERR " NOK: ($stderr)\n";
    }
    # on Windows (MSVC) stdout is set, but not stderr
    else {
      $stdout =~ s/[\r\n]$//;
      $stdout =~ s/.+[\r\n]//;
      $stdout =~ s/^\Q$src\E[\(\)\d\s:]*//;
      print STDERR " NOK: ($stdout)\n";
    }

    return 0;
  }
}

### internal functions
sub _sdl_config_via_script
{
  my $param    = shift;
  my @add_libs = @_;
  my $devnull = File::Spec->devnull();
  my $script = Alien::SDL::ConfigData->config('script');
  return unless ($script && ($param =~ /[a-z0-9_]*/i));
  my $val = `$script --$param 2>$devnull`;
  $val =~ s/[\r\n]*$//;
  if($param eq 'cflags') {
    $val .= ' ' . Alien::SDL::ConfigData->config('additional_cflags');
  }
  elsif($param eq 'libs') {
    $val .= ' ' . join(' ', @add_libs) if scalar @add_libs;
    $val .= ' ' . Alien::SDL::ConfigData->config('additional_libs');
  }
  elsif($param =~ /^(ld_shlib_map|ld_shared_libs|ld_paths)$/) {
    $val = Alien::SDL::ConfigData->config('config')->{$param};
  }
  return $val;
}

sub _sdl_config_via_config_data
{
  my $param    = shift;
  my @add_libs = @_;
  my $share_dir = dist_dir('Alien-SDL');
  my $subdir = Alien::SDL::ConfigData->config('share_subdir');
  return unless $subdir;
  my $real_prefix = catdir($share_dir, $subdir);
  return unless ($param =~ /[a-z0-9_]*/i);
  my $val = Alien::SDL::ConfigData->config('config')->{$param};
  return unless $val;
  # handle additional flags
  if($param eq 'cflags') {
    $val .= ' ' . Alien::SDL::ConfigData->config('additional_cflags');
  }
  elsif($param eq 'libs') {
    $val .= ' ' . join(' ', @add_libs) if scalar @add_libs;
    $val .= ' ' . Alien::SDL::ConfigData->config('additional_libs');
  }
  # handle @PrEfIx@ replacement
  if ($param =~ /^(ld_shared_libs|ld_paths)$/) {
    s/\@PrEfIx\@/$real_prefix/g foreach (@{$val});
  }
  elsif ($param =~ /^(ld_shlib_map)$/) {
    while (my ($k, $v) = each %$val ) {
      $val->{$k} =~ s/\@PrEfIx\@/$real_prefix/g;
    }
  }
  else {
    $val =~ s/\@PrEfIx\@/$real_prefix/g;
  }
  return $val;
}

1;
